src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs

```diff
@@ -686,6 +802,105 @@ void AddReturnValue(MultiValue value)
             }
         }
 
+        private static bool TryGetMakeGenericInstantiation(
+            MethodDesc contextMethod,
+            in MultiValue genericParametersArray,
+            out Instantiation inst,
+            out bool isExact)
+        {
+            // We support calling MakeGeneric APIs with a very concrete instantiation array.
+            // Only the form of `new Type[] { typeof(Foo), typeof(T), typeof(Foo<T>) }` is supported.
+
+            inst = default;
+            isExact = true;
+            Debug.Assert(contextMethod.GetTypicalMethodDefinition() == contextMethod);
+
+            var typesValue = genericParametersArray.AsSingleValue();
+            if (typesValue is NullValue)
+            {
+                // This will fail at runtime but no warning needed
+                inst = Instantiation.Empty;
+                return true;
+            }
+
+            // Is this an array we model?
+            if (typesValue is not ArrayValue array)
+            {
+                return false;
+            }
+
+            int? size = array.Size.AsConstInt();
+            if (size == null)
+            {
+                return false;
+            }
+
+            TypeDesc[]? sigInst = null;
+            TypeDesc[]? defInst = null;
+
+            ArrayBuilder<TypeDesc> result = default;
+            for (int i = 0; i < size.Value; i++)
+            {
+                // Go over each element of the array. If the value is unknown, bail.
+                if (!array.TryGetValueByIndex(i, out MultiValue value))
+                {
+                    return false;
+                }
+
+                var singleValue = value.AsSingleValue();
+
+                TypeDesc? type = singleValue switch
+                {
+                    SystemTypeValue systemType => systemType.RepresentedType.Type,
+                    GenericParameterValue genericParamType => genericParamType.GenericParameter.GenericParameter,
+                    NullableSystemTypeValue nullableSystemType => nullableSystemType.NullableType.Type,
+                    _ => null
+                };
+
+                if (type is null)
+                {
+                    return false;
+                }
+
+                // type is now some type.
+                // Because dataflow analysis oddly operates on method bodies instantiated over
+                // generic parameters (as opposed to instantiated over signature variables)
+                // We need to swap generic parameters (T, U,...) for signature variables (!0, !!1,...).
```

Comment:
I'm not 100% sure, but this might have been to get a behavior similar to what Cecil does. If I remember correctly, Cecil doesn't have a concept of "signature variables", it always instantiates the generics with something parameters instead.
Also - the signature variables are not that interesting for data flow - we would need the generic parameters anyway since that's where the annotations are.

Reply:
It was added in https://github.com/dotnet/runtimelab/pull/354. I'm not proud of it. Analyzing open instantiations is tricky. Yep another option would be to convert everything to definitions before doing any interesting work with it, like Cecil does, to bypass the problem with ambiguous signature variables. But we'd have to do it manually on each use. Not sure if that would be an improvement.

(The problem is that if you want to go over the method signature of a generic method that is itself within a generic method, it becomes ambiguous what the `!!0` refers to, same for methods on generic types. The type system was getting confused and it didn't look straightforward to fix.)